### PR TITLE
Fixes #30344 - use pf4 navigation

### PR DESCRIPTION
**Pull request by MariaAga**
* Continuing https://github.com/theforeman/foreman/pull/8868
* Resolved the bugs described here: https://github.com/theforeman/foreman/pull/8868#issuecomment-950948060
~~* Applied https://community.theforeman.org/t/menu-hiding-delay-poll/32272 in `onPopperMouseLeave`~~
~~* I used `withPopper` instead of the `flyout` prop so we can use `onPopperMouseLeave`~~
* Had to change a lot of css due to pf page css
* `foreman-theme` is used so it will be easier to override `body_css_classes` and change the navigation colors
* Changed some tests from snapshot tests to react testing library
* From what I understand we dont have the capacity to finish the menu re-arrangement so the menu items stay as they are
* Tried to apply all comments from the different discussions but I might have missed some  
* interaction video: [Screencast from 2023-03-22 15-18-57.webm](https://user-images.githubusercontent.com/30431079/226933175-860f425e-881e-4f6e-8f90-4f32e096357d.webm)
* 
image

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/Layout/components/Toolbar/Header.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  dataPropType,
+} from '../../LayoutHelper';
+import HeaderToolbar from './HeaderToolbar';
+
+const Header = ({
+  data: { logo, brand, root, ...props },
+  onNavToggle,
+  isLoading,
+}) => (
+  <PageHeader
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/Layout/components/Toolbar/Header.js]: Page header is being deprecated in favor of masthead so I will change it in this pr

File: webpack/assets/javascripts/react_app/components/Layout/components/LayoutContainer.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ import React, { useEffect } from 'react';
 import PropTypes from 'prop-types';
 
 const LayoutContainer = ({ isCollapsed, children }) => {
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/Layout/components/LayoutContainer.js]: should probably remove this wrapper

File: webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js [hunk 3]
```diff
@@ -20,47 +25,52 @@ const PageLayout = ({
...
-              <div id="toolbar-spinner">
-                <Spinner loading size="sm" />
-              </div>
-            )}
-            {toolbarButtons}
-          </div>
-        </Col>
-      </Row>
-      {children}
+  <div id="react-content">
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js]: also can probably remove this div

File: webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js [hunk 4]
```diff
@@ -20,47 +25,52 @@ const PageLayout = ({
...
+                  <Spinner loading size="sm" />
+                </div>
+              )}
+              {toolbarButtons}
+            </div>
+          </Col>
+        </div>
+      </PageSection>
+    )}
+    <PageSection variant={PageSectionVariants.light}>{children}</PageSection>
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js]: Should have a prop to define page section type

File: app/assets/stylesheets/base.scss [hunk 5]
```diff
@@ -2,13 +2,40 @@ html {
...
+  &.pf-c-page {
+    display: block;
+    grid-template-columns: unset;
+    grid-template-areas: unset;
+  }
 }
 
 body {
   height: 100%;
   width: 100%;
```
Comment (MariaAga) [inline app/assets/stylesheets/base.scss:34]: maybe add 
` overflow-y: auto`
might break long pages like audits

File: webpack/assets/javascripts/react_app/components/Layout/layout.scss [hunk 6]
```diff
@@ -1,146 +1,49 @@
...
-        }
-      }
-    }
-
-    span {
-      color: $nav-pf-vertical-active-icon-color;
-    }
-  }
+#foreman-main-container {
+  height: fit-content;
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/Layout/layout.scss:8]: or height 100% and `#rails-app-content { z-index: 101;`

File: webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js [hunk 7]
```diff
@@ -20,47 +25,52 @@ const PageLayout = ({
...
-        </Col>
-      </Row>
-      {children}
+  <div id="react-content">
+    <Head>
+      <title>{header}</title>
+    </Head>
+    <div id="breadcrumb">
+      {!breadcrumbOptions && (
+        <PageSection variant={PageSectionVariants.light}>
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js]: move `<PageSection` to be above the `<div id="breadcrumb">`
and remove `PageSection` from `Breadcrumb`

File: webpack/assets/javascripts/react_app/components/Layout/layout.scss (hunk 6 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/Layout/layout.scss:8]: #rails-app-content z-index should be more than 101 to show `modal fade` but not too much to not hide `bullet-footer`

File: webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js (hunk 3 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js:28]: or not

**Comment by MariaAga** [comment]
Still not ready and waiting for an answer on how should we add titles in the secondary nav items (like "templates" and "reports")

**Comment by MariaAga** [comment]
design solution:
Screenshot from 2023-04-20 20-42-54

File: app/assets/stylesheets/patternfly_colors_overrides.scss [hunk 8]
```diff
@@ -0,0 +1,28 @@
+@import './colors.scss';
+
+.foreman-theme {
+  .pf-c-page {
+    --pf-c-page--BackgroundColor: var(--pf-global--BackgroundColor--100);
```
Comment (MariaAga) [inline app/assets/stylesheets/patternfly_colors_overrides.scss]: need to move this rule out of foreman-theme as it should always be white

**Comment by sjha4** [comment]
Starting to test this. Looks great! 

One note:
The toggle to expand/minimize nav bar, hides the entire nav bar instead of the current behavior of leaving the icons visible.
Screenshot from 2023-04-24 12-10-16

Screenshot from 2023-04-24 12-13-40

Behavior today:
Screenshot from 2023-04-24 12-12-52

**Comment by MariaAga** [comment]
Unfortunately pf doesnt support only icons as a navigation, we can add a custom icon view if anyone has capacity for it but it wont be soon

**Comment by sjha4** [comment]
Tabs on the Settings page seem to be affected by this PR:

Screenshot from 2023-04-25 16-01-37

**Comment by MariaAga** [comment]
Thanks for the feedback, fixed it with adding `title_filter_parent` css + fixed the ruby tests

**Comment by sjha4** [comment]
Few things:
1. On some pages, mostly angular content page menu items that I see so far, the menu selection doesn't change when switching to that menu item. I think it might be cause angular router doesn't force a page refresh to switch between menu routes.
[screen-capture.webm](https://user-images.githubusercontent.com/21146741/235164366-cac21b20-fa0c-4f06-986c-b7c5b885b3d3.webm)
2. Sticky notifications don't stick to the top bar. Instead there's a gap between top header and the notification. 
Tried to take screenshots to show the difference.
Screenshot from 2023-04-28 09-49-39

-----------------------------------------------------

Screenshot from 2023-04-28 09-47-35

3. On some pages the space between top bar and action drop down and content seems like a lot. One example of this I found was the smart proxy details page:
Screenshot from 2023-05-01 13-18-18

4. There are console warnings about aria-labels:
```
For accessibility reasons an aria-label should be specified on nav groups if a title isn't
at NavGroup (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:46934:11)
    at ul
    at section
    at li
    at NavExpandable (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:46850:9)
    at ul
    at NavList (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:47166:9)
    at nav
    at Nav (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:46763:9)
    at Navigation (https://centos8-katello-devel-stable.example.com:3808/webpack/bundle.js:16033:20)
    at div
    at div
    at PageSidebar (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:49059:11)
    at div
    at Page (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:48508:9)
    at Layout (https://centos8-katello-devel-stable.example.com:3808/webpack/bundle.js:15100:20)
    at ConnectedLayout (https://centos8-katello-devel-stable.example.com:3808/webpack/bundle.js:4656:23)
    at Router (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:671233:30)
    at ConnectedRouter (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:419685:7)
    at ConnectedRouterWithContext (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:419783:25)
    at ConnectFunction (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:669196:75)
    at ApolloProvider (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:5870:21)
    at div
    at ReactApp (https://centos8-katello-devel-stable.example.com:3808/webpack/bundle.js:24439:21)
    at IntlProvider (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:659219:5)
    at I18nProviderWrapper (https://centos8-katello-devel-stable.example.com:3808/webpack/bundle.js:34732:9)
    at Provider (https://centos8-katello-devel-stable.example.com:3808/webpack/foreman-vendor.bundle-v12.0.2-development-dbff1dd3f70d22a3a834.js:668909:20)
    at StoreProvider(I18nProviderWrapper(ReactApp))
    at DataProvider(StoreProvider(I18nProviderWrapper(ReactApp)))
```

**Comment by sjha4** [comment]
Another note: One note regarding sizing of menu items.  Sync status and Lifecycle are at the same level but Lifecycle is displayed smaller. Should they be of the same size since they are both on the same level?

Screenshot from 2023-05-01 12-29-23

**Comment by MariaAga** [comment]
1. fixed
2. not sure what you meant, but I noticed that in some pages the masthead wasnt sticky so fixed that.
for easier screenshots you can run `window.tfm.toastNotifications.notify({message:'tst',  type: 'success', sticky:true});` in the console in any page.
3. reduced pf3 breadcrumbs margin for that space
4,5. removed navgroup to fix both issues

**Comment by MariaAga** [comment]
Bookmarks menu is broken, fixing

**Comment by MariaAga** [comment]
fixed

**Comment by sjha4** [comment]
Couple small things:
1. Spacing seems to have changed on the user dropdown icon-name:
Screenshot from 2023-05-05 14-37-20
Screenshot from 2023-05-05 14-24-28

2. On one content page > Redhat Repositories> The sticky pagination isn't actually sticking to the top and there's a space between header and it for some reason. Feel free to ignore here and file a katello issue if fix is there.
Screenshot from 2023-05-05 14-40-00

**Comment by MariaAga** [comment]
1.Fixed the user icon
2. https://projects.theforeman.org/issues/36367

**Comment by ekohl** [comment]
My biggest worry is that this new layout is a regression in UX and we need to revert it. My reasoning for that while it may work for cloud.redhat.com, Foreman is a different application. Most pages are resource oriented, not workflow oriented. Because of that we have many more navigation items.

A concrete example of that is Puppet. We have a page for Environments, one for Puppet Classes and one for Puppet Class Parameters. Each class parameter belongs to a class. Classes and Environments are also related. That results in 3 links in the navigation. If you would design it workflow based, it would all be integrated in a single page where you can easily navigate to all resources. Similarly we have Hosts. There's All Hosts, Create Host, Register Host and then with Katello there's also Content Hosts.

I often search the menu visually and then a quick interaction (without having to click) helps to work around this. Increasing the steps required to scan through the navigation can hinder users.

We have about a week of development time left in the [Foreman 3.7 development cycle](https://community.theforeman.org/t/foreman-3-7-schedule-and-planning/33077/3). With my release hat on I'm leaning to pushing this to 3.8. If we merge it soon after branching then we have a full release cycle where people using nightly (including developers) get experience with it.

**Comment by MariaAga** [comment]
@ekohl have you seen https://github.com/theforeman/foreman/pull/9681 ?

**Review by sjha4 — DISMISSED** [review]
Thanks @MariaAga for working on this and making all the changes! I am happy with this PR as far as advertised changes go. Tested with foreman + katello box. Leaving a non-maintainer ack post jenkins on this based on my reviews/testing while we wait for foreman maintainers to accept the PR.👍🏼

**Comment by ekohl** [comment]
I had not, but I'm not sure it really addresses what I'm saying.

If we look at the current menu, there are many items that I would restructure. Doing so would address a lot of my concerns.

For example, in Hosts we have (with a vanilla Foreman):

* All Hosts
* Create Host
* Register Host

If we would have a good host overview page then that could a single item. If you look at the number of clicks required, then Hosts -> load hosts page -> do the right workflow (create / register) should simple and if you want the hosts overview then it's even one less step.

Then there's a large sub-block with Provisioning Setup:

* Architectures
* Hardware Models
* Installation Media
* Operating Systems

IMHO this all belongs under Configure more than Hosts. Again, here I'd suggest a single page with all that information. Whether it's 4 tabs on a single page or something more advanced is something you can debate, but if you look at the number of clicks required then it's fairly similar. I'd also argue that Partition Tables and Provisioning Templates under Templates belong under Provisioning Setup, which Provisioning Templates being the most commonly used.

So in short: I think the Foreman application is not ready for the navigation style you propose and a search in the navigation isn't reducing the number of keystrokes/clicks to get to the right page. We can get there, but it's a significant effort.

**Review by sjha4 — COMMENTED** [review]
I am not sure what our support matrix for firefox is but the new menu does not seem styled properly for firefox. Could you confirm if this is also reproducible in your env.
Firefox:
Screenshot from 2023-05-08 14-31-34
Chrome:
Screenshot from 2023-05-08 14-33-05

**Comment by ekohl** [comment]
> I am not sure what our support matrix for firefox is

Should be supported according to https://theforeman.org/manuals/nightly/index.html#3.1.4BrowserCompatibility

**Comment by MariaAga** [comment]
Looks like firefix doesnt handle 100% height as it should, checking for solutions

@ekohl Thanks for the feedback but we already finished the discussion on which design we should use, here: https://community.theforeman.org/t/rfc-navigation-once-again/33065
We also will not be able to refactor the pr easily as there are some features that depend on this prs code, so we will need to revert those as well.
If you want maybe you can continue the navigation items reorganization that started here: https://community.theforeman.org/t/navigation-card-sorting-exercise/27672

**Comment by ekohl** [comment]
> @ekohl Thanks for the feedback but we already finished the discussion on which design we should use, here: https://community.theforeman.org/t/rfc-navigation-once-again/33065

I disagree it was finished. The poll was very suggestive (by not presenting the current option) and the comments in favor of keeping the current navigation were not addressed.

**Comment by MariaAga** [comment]
The design in the poll was changed to include the titles (hosts, provisioning setup, template, etc).
The comments that asked to keep the current navigation also asked to make changes that are not possible without a refactor, and the closes thing in design to the current navigation was Full-length panel flyout, and was not chosen by most people in the poll.
I meant the discussion is finished as some time has passed and the majority have voted for the expendable option.

**Comment by ares** [comment]
TL;DR - Given the current menu has multiple [issues](https://community.theforeman.org/t/menu-hiding-delay-poll/32272/11) that can't be fixed and blocks other features, I think it's better to make a change that majority prefers without further delays.

I personally believe, the navigation has been extensively discussed in the past. Be it the linked [RFC](https://community.theforeman.org/t/rfc-navigation-once-again/33065), or previous discussions [1](https://community.theforeman.org/t/new-pf4-navigation/25838/28) [2](https://community.theforeman.org/t/usabilty-with-main-menu/28709) [3](https://community.theforeman.org/t/new-pf4-navigation/25838). There was also a research performed on face to face meetings with the users recently. While some of the linked discussions didn't cover the expandable option, they showed general feelings about the current menu experience quite well. The solution that majority preferred is the expandable menu. It was also clear, that people had different preferences (not surprising at all) and we can't please all.

Unless I missed something, the UX concerns raised in this PR were raised here for the first time (number of clicks this brings), quite late in the process. It's not something that would prevent anyone completely from using the application. I believe the direction was already approved by users who care for upcoming changes, and the PR can be merged, provided there are no more technical issues like firefox support. If I understand correctly, the usability concerns do not endanger the 3.7 release stability, @ekohl please correct me if I'm wrong.

That does not mean, we wouldn't receive tons of negative feedback when we release 3.7. If that happened, we could work on a new implementation. At that moment, we would gather the feedback and re-evaluate, but currently we have very few inputs like this. In any case, we would of course continue to improve the implementation. I believe it's better to move and iterate on improvements, than stay still. Especially if this blocks other two improvements.

I also believe that the discussion on further improvements or alternatives does not belong to here. The scope of this PR is given, we can talk further on discourse or our IRC channel, where the original discussions happened and where the user community is present. I agree we should have communicated better, that we can not fix the delays with the current implementation and how we're going to address the scrolling concern, let's fix that.

If people with access to the demo hw wants to get hands on experience, I just deployed the patch from this PR there. I find it pretty good after short time playing with it (it's important to navigate to the page to see how menu expands afterward).

**Comment by ares** [comment]
this gap between the menu and header caught my eye (chrome 113), it wasn't there before and IMHO does not look nice, is that something that can be fixed in this PR?
Screenshot_20230510_163247

EDIT: reproducing steps - Unflold Monitor, then Fold it again

**Comment by MariaAga** [comment]
This gap is on purpose, see here: https://www.patternfly.org/v4/components/navigation/react-demos/default-nav/

Currently working on fixing the firefox issue, and there is a small issue of rails tooltips in select all hosts being hidden by the nav.
Since there are a lot of small issues like these discovered there are probably more undiscovered, and since we cant add the search in nav in time for 3.7 I think it makes sense to wait for 3.8

**Comment by ares** [comment]
Also, the header does not have the user menu aligned to the right anymore, I didn't see that in your screenshots, is that something broken on my side?

masthead

I see the following error in the console

```
Scrollbar test exception: TypeError: Cannot read properties of null (reading 'appendChild') browser.js:69
```

**Comment by ares** [comment]
> Currently working on fixing the firefox issue, and there is a small issue of rails tooltips in select all hosts being hidden by the nav.
> Since there are a lot of small issues like these discovered there are probably more undiscovered, and since we cant add the search in nav in time for 3.7 I think it makes sense to wait for 3.8

Ok, if there are issues we won't be able to resolve in time for 3.7, then it will have to wait.

**Comment by ekohl** [comment]
I'll say that with my user hat on I'm skeptical of the change, but it's mostly my maintainer hat that's worried about the potential issues. As @MariaAga said, there are already known issues but as always with big changes probably also a lot of undiscovered ones. The Foreman 3.7 development cycle ends this week, after which we have stabilization week. I'd feel a lot more confident in the best outcome if we don't rush this in at the last moment but rather merge it at the start of the 3.8 release cycle, giving us plenty of time for feedback.

**Comment by MariaAga** [comment]
* Scrollbar error should be fixed here: https://github.com/theforeman/foreman/pull/9697/
* user alignment: I hope its just an issue of not running `assets:clobber` or refreshing, as I didnt see this in my envs

**Comment by MariaAga** [comment]
Fixed the firefox height issue, looking into the test failure

**Review by Ron-Lavi — COMMENTED** [review]
great work @MariaAga ! seems like you put a lot of effort into this.
code wise it looks good to me,
for styling, I had two thoughts:
- Can we clear the padding at the top of the menu:
Screenshot - 2023-05-21T115151 697

-  Can we increase a bit the font size of the sub-items? feels somewhat hidden, but maybe it's just me
Screenshot - 2023-05-21T122029 315

**Review by Ron-Lavi — COMMENTED** [review]
Looks like the transition between two react pages slowed down, I wonder why:

- develop:
[Screencast from 2023-05-21 12-32-59.webm](https://github.com/theforeman/foreman/assets/26363699/bb572555-9bcd-40ed-8616-6059540ca482)

- this branch:
[Screencast from 2023-05-21 12-36-09.webm](https://github.com/theforeman/foreman/assets/26363699/94e5c1b1-8e75-46eb-8297-cca3e9409ef1)

in addition, there needs to be some indication that the user clicked on the nav item and something is loading.

**Comment by MariaAga** [comment]
Thanks for the review! @Ron-Lavi 
 
> Can we clear the padding at the top of the menu:

we shouldnt: https://github.com/theforeman/foreman/pull/9659#issuecomment-1542332003

> Can we increase a bit the font size of the sub-items? feels somewhat hidden, but maybe it's just me

Size is defined by pf: 
https://www.patternfly.org/v4/components/navigation#expandable-third-level

>Looks like the transition between two react pages slowed down, I wonder why:

didn't use `navigate`, so it loaded everything on change page instead of just the new react. fixed now

> in addition, there needs to be some indication that the user clicked on the nav item and something is loading.

Do you have an example somewhere? I didnt see anything like that in old navigation or in pf docs

also removed unused `getActiveMenuItem` and fixed active menu not changing on browser history change

**Comment by Ron-Lavi** [comment]
> didn't use `navigate`, so it loaded everything on change page instead of just the new react. fixed now

awesome! (didn't test yet)

> Do you have an example somewhere? I didnt see anything like that in old navigation or in pf docs

I think there was some spinner to indicate page transition, maybe this one? 
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/LayoutActions.js#LL27C14-L27C25

**Comment by MariaAga** [comment]
`showLoading` is `showSpinner`
The spinner was used for jquery actions, like:
``` jquery
function os_selected(element) {
....
  tfm.tools.showSpinner();
```
And is used in this pr here: `<Header data={data} onNavToggle={onNavToggle} isLoading={isLoading} />`

**Comment by sjha4** [comment]
Tested the firefox fix and it looks nice now.. 👍🏼

**Review by Ron-Lavi — COMMENTED** [review]
@MariaAga code looks good, can you rebase?

**Comment by MariaAga** [comment]
Thanks, rebased

**Review by Ron-Lavi — APPROVED** [review]
Tested and it works great, merging,
Thanks @MariaAga !
